File: services/api/app/models/feedback.py
from typing import Any, Literal
from pydantic import Field

from app.models.domain import CamelModel


ReportType = Literal["bug", "feature", "ux", "general"]
ProductArea = Literal["command_center", "profile", "community", "calendar", "lookup", "other"]


class CreateFeedbackRequest(CamelModel):
    report_type: ReportType
    product_area: ProductArea
    title: str = Field(min_length=3, max_length=120)
    description: str = Field(min_length=10, max_length=4000)
    expected_behavior: str | None = Field(default=None, max_length=2000)
    page_path: str | None = Field(default=None, max_length=500)
    user_agent: str | None = Field(default=None, max_length=1000)
    metadata: dict[str, Any] = Field(default_factory=dict)


class FeedbackSubmissionOut(CamelModel):
    id: str
    created_at: str

File: frontend/src/lib/api/feedback.ts
import { createClient } from "@/lib/supabase/client";
import { getApiBaseUrl } from "@/lib/api/client";

export type FeedbackReportType = "bug" | "feature" | "ux" | "general";
export type FeedbackProductArea =
  | "command_center"
  | "profile"
  | "community"
  | "calendar"
  | "lookup"
  | "other";

export type SubmitFeedbackPayload = {
  reportType: FeedbackReportType;
  productArea: FeedbackProductArea;
  title: string;
  description: string;
  expectedBehavior?: string | null;
  pagePath?: string | null;
  userAgent?: string | null;
  metadata?: Record<string, unknown>;
};

export type FeedbackSubmission = {
  id: string;
  createdAt: string;
};

async function getAccessToken(): Promise<string> {
  const supabase = createClient();
  const {
    data: { session },
  } = await supabase.auth.getSession();
  if (!session?.access_token) throw new Error("Not authenticated");
  return session.access_token;
}

async function readApiErrorMessage(res: Response, fallback: string): Promise<string> {
  const detail = await res.json().catch(() => null);
  const payload = detail?.detail ?? detail;
  if (typeof payload === "string") return payload;
  if (payload && typeof payload === "object") {
    if (typeof payload.message === "string") return payload.message;
    if (typeof payload.detail === "string") return payload.detail;
  }
  return fallback;
}

export async function submitFeedback(payload: SubmitFeedbackPayload): Promise<FeedbackSubmission> {
  const token = await getAccessToken();
  const res = await fetch(`${getApiBaseUrl()}/api/feedback`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${token}`,
    },
    body: JSON.stringify(payload),
  });
  if (!res.ok) {
    throw new Error(await readApiErrorMessage(res, `submitFeedback failed: ${res.status}`));
  }
  return res.json() as Promise<FeedbackSubmission>;
}
